### PR TITLE
[MOSIP-30646] Automated ssl certificates renewal.

### PR DESCRIPTION
**Pull request by Mohanraj209**

### CODE REVIEW

File: utils/ssl_renewel_util/certs_renew_cron.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+cron_time="0 6 $day $month_num *"
+
+# Print the cron time (for verification)
+echo "Cron time: $cron_time"
+
+# Define the command you want to run on the expiry date
+command_to_run="renew_certificates.sh"
+
+# Create a cron job
+(crontab -l; echo "$cron_time $command_to_run") | crontab -
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/certs_renew_cron.sh:43]: print crontab after adding it 

```
$ echo "0 0 1 1 0 2 xxxxxx" | crontab -
techno-384@techno384-Latitude-3410:~$ crontab -l
0 0 1 1 0 2 xxxxxx
```

File: utils/ssl_renewel_util/certs_renew_cron.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+esac
+
+# Construct the cron time string
+cron_time="0 6 $day $month_num *"
+
+# Print the cron time (for verification)
+echo "Cron time: $cron_time"
+
+# Define the command you want to run on the expiry date
+command_to_run="renew_certificates.sh"
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/certs_renew_cron.sh:40]: please remove this variable

File: utils/ssl_renewel_util/certs_renew_cron.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+# Prompt the user for the domain name
+read -p "Enter the domain name: " mosip_domain
+
+expiry_date=$(sudo openssl x509 -enddate -noout -in /etc/letsencrypt/live/$mosip_domain/fullchain.pem)
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/certs_renew_cron.sh:6]: update openssl command

```
date --date="$(sudo openssl x509 -enddate -noout -in /etc/letsencrypt/live/dev.mosip.net/fullchain.pem | sed -e 's/notAfter=//g') -7 days" +'%d %m'
```

renew ssl certs before 7 day of expiry

File: utils/ssl_renewel_util/certs_renew_cron.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  Jun) month_num=6 ;;
+  Jul) month_num=7 ;;
+  Aug) month_num=8 ;;
+  Sep) month_num=9 ;;
+  Oct) month_num=10 ;;
+  Nov) month_num=11 ;;
+  Dec) month_num=12 ;;
+  *) echo "Invalid month"; exit 1 ;;
+esac
+
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/certs_renew_cron.sh:32]: please remove this as it is not required

File: utils/ssl_renewel_util/renew_ssl_certs.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+# Script to renew ssl certificates..
+
+# Prompt the user for the existing ssl certs expiry date for backing up the Dir with specific date.
+read -p "Enter the expiry date of existing ssl certs (e.g., 2024-12-31): " expiry_date
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/renew_ssl_certs.sh:5]: please remove this

File: utils/ssl_renewel_util/renew_ssl_certs.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+# Define the source and destination directories
+src_dir="/etc/letsencrypt/live"
+dest_dir="/etc/letsencrypt/live-${expiry_date}"
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/renew_ssl_certs.sh:9]: use `date +"%d-%m-%Y"` instead

File: utils/ssl_renewel_util/renew_ssl_certs.sh [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    # Move the directory
+    sudo mv "$src_dir" "$dest_dir"
+    echo "Directory moved to $dest_dir"
+else
+    echo "Source directory $src_dir does not exist."
+fi
+
+# Regenerate the ssl certificates
+# Prompt the user for the domain name
+read -p "Enter the domain name (e.g., sandbox.xyz.net): " mosip_domain
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/renew_ssl_certs.sh:22]: check for the existence of `mosip_domain`, if not found prompt it

Comment (syedsalman3753) [reply utils/ssl_renewel_util/renew_ssl_certs.sh:22]: for ssl-renew logger:

```
sudo mv /tmp/ssl-renew.log /tmp/ssl-renew.log.old || true
LOG_FILE="/tmp/ssl-renew.log"
ENV_FILE_PATH="/etc/environment"
source $ENV_FILE_PATH
```

File: utils/ssl_renewel_util/renew_ssl_certs.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+if [ -z "$mosip_domain" ]; then
+    echo "Domain name cannot be empty"
+    exit 1
+fi
+
+# Run certbot with the user-provided domain name
+sudo certbot certonly \
+  --dns-route53 \
+  -d "$mosip_domain" \
+  -d "*.$mosip_domain"
```
Comment (syedsalman3753) [inline utils/ssl_renewel_util/renew_ssl_certs.sh:34]:
```
sudo certbot certonly --dns-route53 -d "*.${mosip_domain}" -d "${mosip_domain}" --non-interactive --agree-tos --email "$certbot_email"
```

**Review by syedsalman3753 — CHANGES_REQUESTED** [review]

File: utils/ssl_renewel_util/README.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
+# SSL Certificate Renewal Script for Aws
+Note: If you are not using Aws as your Dns provider then this script won't work. You'll have to manually renewel the certificates, In case if you are using any other Dns provider then check if certbot supports any dns plugin for your Dns provider and then just update the certbot command then the script should work fine.
```
Comment (ckm007) [inline utils/ssl_renewel_util/README.md:2]: If you are not using Aws --> If you are not using Route 53 from Aws

File: utils/ssl_renewel_util/certs_renew_cron.sh (hunk 1 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/certs_renew_cron.sh:43]: updated

File: utils/ssl_renewel_util/README.md (hunk 9 above)
Comment (ckm007) [reply utils/ssl_renewel_util/README.md:2]: renewel --> renew

File: utils/ssl_renewel_util/README.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+This script is used for the renewal of SSL certificates using Let's Encrypt's `certbot`. It moves old certificates to a backup directory with an expiry date and renew the ssl certificates and Update cert location.
+
+## Prerequisites
+
+- **Certbot**: Ensure `certbot` is installed and configured on your system.
+- **AWS Route 53**: The script uses Route 53 for DNS challenges.
+- **Sudo Privileges**: Required to move directories and restart services.
+- **Nginx**: Ensure Nginx is installed and configured to use Let's Encrypt certificates.
```
Comment (ckm007) [inline utils/ssl_renewel_util/README.md:11]: also add a line informing e assume your private and public sertificate is assumed to be present in these respective directpries.

File: utils/ssl_renewel_util/README.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+## Installation
+
+1. **Install Certbot**: Follow the instructions [here](https://certbot.eff.org/instructions) to install `certbot` for your nginx server and operating system.
```
Comment (ckm007) [inline utils/ssl_renewel_util/README.md:15]: **Install Certbot**: Follow the instructions [here](https://certbot.eff.org/instructions) to install `certbot` for your nginx server

File: utils/ssl_renewel_util/README.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+2. **Configure AWS Route 53**: Ensure your domain's DNS is managed by AWS Route 53 and that you have the necessary permissions to perform DNS validation.
```
Comment (ckm007) [inline utils/ssl_renewel_util/README.md:16]: Also AWS credentials are already set in .aws dir.

File: utils/ssl_renewel_util/certs_renew_cron.sh (hunk 2 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/certs_renew_cron.sh:40]: updated

**Review by ckm007 — CHANGES_REQUESTED** [review]

File: utils/ssl_renewel_util/certs_renew_cron.sh (hunk 4 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/certs_renew_cron.sh:32]: updated

File: utils/ssl_renewel_util/renew_ssl_certs.sh (hunk 6 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/renew_ssl_certs.sh:9]: updated

File: utils/ssl_renewel_util/certs_renew_cron.sh (hunk 3 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/certs_renew_cron.sh:6]: updated

File: utils/ssl_renewel_util/renew_ssl_certs.sh (hunk 8 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/renew_ssl_certs.sh:34]: updated

File: utils/ssl_renewel_util/README.md (hunk 9 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/README.md:2]: updated

File: utils/ssl_renewel_util/README.md (hunk 10 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/README.md:11]: updated

File: utils/ssl_renewel_util/README.md (hunk 12 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/README.md:16]: updated

File: utils/ssl_renewel_util/README.md (hunk 11 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/README.md:15]: updated

File: utils/ssl_renewel_util/renew_ssl_certs.sh (hunk 7 above)
Comment (Mohanraj209) [reply utils/ssl_renewel_util/renew_ssl_certs.sh:22]: updated

File: utils/ssl_renewel_util/README.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
+# SSL Certificate Renewal Script for Aws
+Note: If you are not using Route 53 from Aws as your Dns provider then this script won't work. You'll have to manually renewel the certificates, In case if you are using any other Dns provider then check if certbot supports any dns plugin for your Dns provider and then just update the certbot command within the script then the `renew_ssl_certs` script should work fine.
```
[the remark this answers is not in the document]
Comment (Mohanraj209) [reply utils/ssl_renewel_util/README.md]: updated

**Review by ckm007 — APPROVED** [review]

**Pull request merged**